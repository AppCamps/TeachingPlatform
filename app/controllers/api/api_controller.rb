# frozen_string_literal: true

module Api
  class ApiController < ::ApplicationController
    AuthenticationError = Class.new(StandardError)
    AuthenticationMissingError = Class.new(AuthenticationError)
    AuthenticationPrivacyPolicyMissingError = Class.new(AuthenticationError)

    around_action :with_english_locale

    before_action :require_json_api_mime_type
    before_action :authenticate
    before_action :require_accepted_privacy_policy
    before_action :deep_underscore_param_keys

    after_action :set_content_type

    skip_before_action :authenticate, only: :handle_not_found
    skip_before_action :require_accepted_privacy_policy, only: :handle_not_found

    rescue_from AuthenticationError,
                with: :handle_authentication_error
    rescue_from AuthenticationMissingError,
                with: :handle_authentication_missing_error
    rescue_from ActionController::ParameterMissing,
                with: :handle_action_controller_parameter_missing
    rescue_from ActiveRecord::RecordNotFound,
                with: :handle_not_found

    def handle_not_found
      head :not_found
    end

    protected

    def cached(key)
      return yield unless key

      Rails.cache.fetch(key) do
        yield
      end
    end

    def serialize_json(resource, serializer_options: {}, adapter_options: {})
      ::JsonApiSerializerService.new(resource).as_json(
        serializer_options: {
          scope: current_user,
          scope_name: :current_user
        }.merge(serializer_options),
        adapter_options: adapter_options
      )
    end

    private

    def with_english_locale
      I18n.with_locale(:en) do
        yield
      end
    end

    def transform_key(key)
      key.to_s.underscore
    end

    def deep_underscore_param_keys
      # TODO: try to come up with something better...
      params.send(:parameters).send(:try, :deep_transform_keys!, &:underscore)
    end

    def require_json_api_mime_type
      head(:not_acceptable) unless request.content_type == Mime[:jsonapi].to_s
    end

    def handle_action_controller_parameter_missing(exception)
      Rollbar.error(exception)

      klass = Class.new
      klass.include ActiveModel::Validations

      resource = klass.new
      resource.errors.add(exception.param, exception.message)

      render json: resource, status: :bad_request,
             serializer: ActiveModel::Serializer::ErrorSerializer
    end

    def handle_authentication_error(error)
      message = error.message if error
      error = { code: 'authentication_error', title: message }
      render json: { errors: [error] }, status: :unauthorized
    end

    def handle_authentication_missing_error(_error)
      error = { code: 'authentication_error' }
      render json: { errors: [error] }, status: :unauthorized
    end

    def authenticate
      warden.authenticate(scope: :user)
      raise AuthenticationMissingError, nil if current_user.blank?
    end

    def require_accepted_privacy_policy
      return if !current_user || current_user.privacy_policy_accepted
      raise AuthenticationPrivacyPolicyMissingError, 'Privacy Policy not accepted'
    end

    def set_content_type
      self.content_type = Mime[:jsonapi].to_s
    end
  end
end
