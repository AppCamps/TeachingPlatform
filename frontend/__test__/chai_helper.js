import chai, { expect } from "chai";
import chaiEnzyme from "chai-enzyme";

chai.use(chaiEnzyme());
chai.config.truncateThreshold = 0;
chai.config.showDiff = true;

export default chai;

export { expect };
