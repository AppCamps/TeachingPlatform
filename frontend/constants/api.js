export const API_FETCHED = 'API_FETCHED';
