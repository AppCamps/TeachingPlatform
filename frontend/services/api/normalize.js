import camelize from "camelize";

function normalizeRecord(data) {
  const relationships = {};
  if (data.relationships && data.relationships !== []) {
    Object.keys(data.relationships).forEach((key) => {
      const relationshipData = data.relationships[key].data;
      if (Array.isArray(relationshipData)) {
        relationships[key] = relationshipData.map((hash) => hash.id);
      } else {
        if (relationshipData === null) {
          return;
        }
        relationships[key] = relationshipData.id;
      }
    });
  }
  const record = {
    id: data.id,
    ...data.attributes,
    ...relationships,
    ...data.meta,
  };
  return {
    type: camelize(data.type),
    record,
  };
}

export function normalizeError(data) {
  if (!data.errors) {
    throw Error(data);
  }

  return data.errors.reduce((memo, error) => {
    if (error.source && error.source.pointer) {
      const key = camelize(
        error.source.pointer.replace(/\/data\/attributes\//, "")
      );
      memo[key] = error.detail;
    }
    return memo;
  }, {});
}

export function normalize(result) {
  let records = [];
  if (Array.isArray(result.data)) {
    records = records.concat(result.data.map((data) => normalizeRecord(data)));
  } else {
    records.push(normalizeRecord(result.data));
  }
  if (result.included) {
    records = records.concat(
      result.included.map((data) => normalizeRecord(data))
    );
  }

  return records.reduce((memo, { type, record }) => {
    if (!memo[type]) {
      memo[type] = {};
    }
    memo[type][record.id] = camelize(record);
    return memo;
  }, {});
}

export function sessionWithUser(result) {
  return {
    expireAfter: result.data.attributes["expire-after"] * 1000,
    session: {
      token: result.data.attributes.token,
    },
    user: {
      id: result.included[0].id,
      ...camelize(result.included[0].attributes),
    },
  };
}
