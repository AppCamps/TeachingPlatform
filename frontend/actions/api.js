import "url-search-params-polyfill";

import { API_FETCHED } from "../constants/api";

export function apiFetched(payload) {
  return {
    type: API_FETCHED,
    payload,
  };
}
