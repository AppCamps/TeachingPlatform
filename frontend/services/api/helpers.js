import 'es6-promise/auto';

import axios from 'axios';

import { SubmissionError } from 'redux-form';

import { log, trackError } from '../../debug';
import { t } from '../../utils/translations';
import { notifications } from '../../config';

import { requestNotification } from '../../actions/notifications';

import { normalize as normalizeFn, normalizeError as normalizeErrorFn } from './normalize';

axios.defaults.xsrfHeaderName = 'X-CSRF-TOKEN';

let dispatch = null;
export function initializeApi(store) {
  dispatch = store.dispatch;
}

function prependLog(url, method, type) {
  return `${method} ${url}: [${type}]`;
}

function logResponse(response) {
  const { headers, config } = response;
  const { url, method } = config;

  if (!headers) {
    return;
  }

  const requestId = headers['X-Request-Id'] || headers['x-request-id'];
  log(
    `${prependLog(
      url,
      method.toUpperCase(),
      'Response',
    )} Statuscode ${response.status} (Request ID: ${requestId})`,
  );
}

function handleSuccess(result) {
  logResponse(result);
  return Promise.resolve(result.data);
}

function handleError(result) {
  const { response } = result;
  if (response && response.status && response.status < 500) {
    logResponse(response);
    if (response.status === 429) {
      dispatch(
        requestNotification({
          type: notifications.failure,
          text: t('Too many requests. Please try again in a few seconds.'),
        }),
      );
    }
    return Promise.reject(response.data);
  }
  trackError(result);
  dispatch(
    requestNotification({
      type: notifications.failure,
      text: t(
        'There was an Error processing your request. Please contact philipp@appcamps.de if the problem persists.',
      ),
    }),
  );
  throw Error(result);
}

export function fetch(endpoint, description) {
  if (!dispatch) {
    throw new Error('Api service has not been initialized with store');
  }

  description.url = endpoint;
  description.headers = {
    'Content-Type': 'application/vnd.api+json',
    Accept: 'application/vnd.api+json',
    ...(description.headers || {}),
  };
  // send {} as data otherwise axios omits the Content-Type header
  if (!description.data) {
    description.data = {};
  }

  log(`${prependLog(endpoint, description.method, 'Request')}`);

  return axios
    .request(description)
    .then(handleSuccess)
    .catch(handleError);
}

export function normalize(customNormalizeFunction) {
  return (result) => {
    if (customNormalizeFunction) {
      return customNormalizeFunction(result);
    }
    return normalizeFn(result);
  };
}

export function normalizeErrors(customNormalizeErrorFn, throwSubmissionError = false) {
  return (response) => {
    if (customNormalizeErrorFn) {
      return Promise.reject(customNormalizeErrorFn(response));
    }
    const errors = normalizeErrorFn(response);
    if (throwSubmissionError) {
      throw new SubmissionError(errors);
    }
    return Promise.reject(errors);
  };
}
