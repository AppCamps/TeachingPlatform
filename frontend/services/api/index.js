import { initializeApi, fetch, normalize, normalizeErrors } from "./helpers";

import SessionSerializer from "../../serializers/session";
import UserSerializer from "../../serializers/user";
import UserRegistrationSerializer from "../../serializers/user/registration";
import LocalitySerializer from "../../serializers/locality";

// api has to be initialized with store in order to dispatch errors
export { initializeApi };

function legacyError(result) {
  return result.errors;
}

export function createUser(user) {
  const payload = UserRegistrationSerializer.serialize(user);

  return fetch("/api/user", {
    method: "POST",
    data: JSON.stringify(payload),
  })
    .then(normalize())
    .catch(normalizeErrors(null, true));
}

export function createSession(loginData) {
  const payload = SessionSerializer.serialize({
    email: loginData.email,
    password: loginData.password,
  });

  return fetch("/api/session", {
    method: "POST",
    data: JSON.stringify(payload),
  })
    .then(normalize())
    .catch(normalizeErrors(legacyError));
}

export function getSession() {
  return fetch("/api/session", {
    method: "GET",
  })
    .then(normalize())
    .catch(normalizeErrors(legacyError));
}

export function deleteSession() {
  return fetch("/api/session", {
    method: "DELETE",
  })
    .then(normalize())
    .catch(normalizeErrors());
}

export function getCourses() {
  return fetch("/api/courses", {
    method: "GET",
  }).then(normalize());
}

export function updateCourseSchoolClass(payload) {
  return fetch(`/api/course_school_classes/${payload.data.id}`, {
    method: "PUT",
    data: JSON.stringify(payload),
  })
    .then(normalize())
    .catch(normalizeErrors(null, true));
}

export function getPreparationMaterials() {
  return fetch("/api/preparation_materials", {
    method: "GET",
  }).then(normalize());
}

export function updateUser(user) {
  const payload = UserSerializer.serialize(user);

  return fetch("/api/user", {
    method: "PUT",
    data: JSON.stringify(payload),
  })
    .then(normalize())
    .catch(normalizeErrors(null, true));
}

export function getClasses() {
  return fetch("/api/classes", {
    method: "GET",
  }).then(normalize());
}

export function getClassById(id) {
  return fetch(`/api/classes?filter[classes.id]=${id}`, {
    method: "GET",
  }).then(normalize());
}

export function createClass(payload) {
  delete payload.data.id;

  return fetch("/api/classes", {
    method: "POST",
    data: JSON.stringify(payload),
  }).then(normalize());
}

export function updateClass(klassId, payload) {
  return fetch(`/api/classes/${klassId}`, {
    method: "PUT",
    data: JSON.stringify(payload),
  }).then(normalize());
}

export function archiveClass(klassId) {
  return fetch(`/api/classes/${klassId}/archive`, {
    method: "PUT",
  }).then(normalize());
}

export function updateCompletedLessonsRelation(classId, payload) {
  return fetch(`/api/classes/${classId}/relationships/completed-lessons`, {
    method: "PUT",
    data: JSON.stringify(payload),
  }).then(normalize());
}

export function deleteCompletedLessonsRelation(classId, payload) {
  return fetch(`/api/classes/${classId}/relationships/completed-lessons`, {
    method: "DELETE",
    data: JSON.stringify(payload),
  }).then(normalize());
}

export function getCountries() {
  return fetch("/api/countries", {
    method: "GET",
  }).then(normalize());
}

export function getPosts(fragments) {
  let url = `/api/posts?page[number]=${fragments.number || 1}`;
  if (fragments.size) url += `&page[size]=${fragments.size}`;

  return fetch(url, {
    method: "GET",
  }).then((result) => ({
    data: normalize()(result),
    links: result.links,
  }));
}

export function createLocality(locality) {
  const payload = LocalitySerializer.serialize(locality);

  return fetch("/api/locality", {
    method: "POST",
    data: JSON.stringify(payload),
  })
    .then(normalize())
    .catch(normalizeErrors(null, true));
}

export function passwordResetRequest(formData) {
  const payload = UserRegistrationSerializer.serialize(formData);

  return fetch("/api/password-reset", {
    method: "POST",
    data: JSON.stringify(payload),
  });
}

export function passwordReset(formData, passwordResetToken) {
  const payload = UserRegistrationSerializer.serialize(formData);
  return fetch(`/api/password-reset/${passwordResetToken}`, {
    method: "PUT",
    data: JSON.stringify(payload),
  })
    .then(normalize(() => ({})))
    .catch(normalizeErrors(null, true));
}

export function emailConfirmationRequest(formData) {
  const payload = UserRegistrationSerializer.serialize(formData);

  return fetch("/api/confirmations", {
    method: "POST",
    data: JSON.stringify(payload),
  });
}

export function emailConfirmation(emailConfirmationToken) {
  const payload = { data: {} };
  return fetch(`/api/confirmations/${emailConfirmationToken}`, {
    method: "PUT",
    data: JSON.stringify(payload),
  })
    .then(normalize(() => ({})))
    .catch(normalizeErrors(null, true));
}
