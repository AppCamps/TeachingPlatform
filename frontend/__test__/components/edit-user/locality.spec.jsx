import React from 'react';
import PropTypes from 'prop-types';

import { shallow, mount } from 'enzyme';

import { expect } from '../../chai_helper';

import EditLocality, { __RewireAPI__ } from '../../../components/edit-user/locality';

import { user } from '../../fixtures/api/createUser';

function LocalityForm() { return <form />; }

describe('<EditLocality />', () => {
  beforeEach(() => {
    __RewireAPI__.__Rewire__('LocalityForm', LocalityForm);
  });

  it('renders includes', () => {
    const wrapper = shallow(
      <EditLocality user={user} />,
      { context: { t: t => t } },
      { childContextTypes: { t: PropTypes.func } },
    );
    global.expect(wrapper).toMatchSnapshot();
  });

  it('includes LocalityForm and forwards all props', () => {
    const wrapper = mount(
      <EditLocality user={user} />,
      { context: { t: t => t } },
      { childContextTypes: { t: PropTypes.func } },
    );
    expect(wrapper.find(LocalityForm)).to.have.props(wrapper.props());
  });
});
