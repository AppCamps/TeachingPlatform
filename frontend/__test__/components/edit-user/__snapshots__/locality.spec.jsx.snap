// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditLocality /> renders includes 1`] = `
<div
  className="container"
>
  <div
    className="editUserPassword"
  >
    <h1
      className="heading"
    >
      Edit your data
    </h1>
    <div
      className="cancel"
    >
      <LinkWrapper
        button={false}
        className={null}
        href={null}
        isAction={false}
        isGreen={false}
        isSecondary={false}
        leftIcon={null}
        rightIcon={null}
        to="/"
      >
        cancel
      </LinkWrapper>
    </div>
    <div
      className="editUserPasswordForm"
    >
      <h2>
        Change address
      </h2>
      <LocalityForm
        user={
          Object {
            "email": "test1@appcamps.de",
            "firstName": "Test",
            "lastName": "User",
            "password": "testtest",
            "passwordConfirmation": "testtest",
            "privacyPolicyAccepted": true,
            "referal": "test123",
            "role": "role_teacher",
          }
        }
      />
    </div>
  </div>
</div>
`;
